src/main/java/org/wso2/carbon/module/csv/CsvToJsonTransformer.java

```diff
@@ -0,0 +1,133 @@
+    void mediate(SimpleMessageContext mc, Stream<String[]> csvArrayStream, String[] header) {
+
+        final EmptyCsvValueType treatEmptyCsvValueAs =
+                getEnumParam(mc, ParameterKey.CSV_EMPTY_VALUES, EmptyCsvValueType.class, EmptyCsvValueType.NULL);
+        final Optional<String> jsonKeysQuery = getStringParam(mc, ParameterKey.JSON_KEYS);
+        final List<JsonDataTypesSchema> dataTypesSchemaList =
+                getJsonArrayParam(mc, ParameterKey.DATA_TYPES, JsonDataTypesSchema.class);
+        final Optional<String> rootJsonKeyQuery = getStringParam(mc, ParameterKey.ROOT_JSON_KEY);
+
+        String[] jsonKeys = generateObjectKeys(jsonKeysQuery.orElse(""), header);
+        Map<Integer, String> dataTypesMap = getDataTypes(dataTypesSchemaList, header);
+
+        csvArrayStream
+                .map(row -> {
+                    JsonObject jsonObject = new JsonObject();
+
+                    for (int i = 0; i < row.length; i++) {
+                        String dataTypeString = dataTypesMap.getOrDefault(i, JsonDataType.STRING.toString());
+                        JsonPrimitive value = getCellValue(row, i, treatEmptyCsvValueAs, dataTypeString);
+                        String key = getObjectKey(jsonKeys, i);
+                        jsonObject.add(key, value);
+                    }
+
+                    return jsonObject;
+                })
+                .collect(rootJsonKeyQuery.map(mc::collectToJsonArray).orElseGet(mc::collectToJsonArray));
+    }
+
+    private JsonPrimitive getCellValue(String[] row, int index, EmptyCsvValueType emptyCsvValueType,
+                                       String dataTypeString) {
+
+        JsonPrimitive cellValue = null;
+        String cellValueString = row[index];
+
+        if (StringUtils.isNotBlank(cellValueString)) {
+            cellValue = convertCellType(cellValueString, dataTypeString);
+        } else if (emptyCsvValueType == EmptyCsvValueType.EMPTY) {
+            cellValue = new JsonPrimitive("");
+        }
+
+        return cellValue;
+    }
+
+    private JsonPrimitive convertCellType(String cellValueString, String dataTypeString) {
+
+        if (dataTypeString != null) {
+            try {
+                JsonDataType dataType = JsonDataType.valueOf(dataTypeString.trim().toUpperCase());
+                switch (dataType) {
+                    case NUMBER:
+                        return new JsonPrimitive(Double.parseDouble(cellValueString));
+                    case INTEGER:
+                        return new JsonPrimitive(Integer.parseInt(cellValueString));
+                    case BOOLEAN:
+                        return new JsonPrimitive(Boolean.parseBoolean(cellValueString));
+                    case STRING:
+                    default:
+                        return new JsonPrimitive(cellValueString);
+                }
+
+            } catch (Exception e) {
+                log.debug("Error converting csv data to given json type : ", e);
```

Comment:
Shall we catch the specific exception and log it as an error instead of a debug and handle it accordingly?

Reply:
foxed it